lib/search/embeddings/embedding.ex

```diff
@@ -0,0 +1,137 @@
+defmodule Search.Embeddings.Embedding do
+  defmacro __using__(opts \\ []) do
+    %{
+      model: model_repo,
+      embedding_size: embedding_size,
+      serving_opts: serving_opts,
+      load_model_opts: load_model_opts,
+      load_tokenizer_opts: load_tokenizer_opts
+    } =
+      opts
+      |> Keyword.validate!([
+        :model,
+        :embedding_size,
+        serving_opts: [],
+        load_model_opts: [],
+        load_tokenizer_opts: []
+      ])
+      |> Map.new()
+
+    batch_size =
+      serving_opts
+      |> Keyword.get(:compile, [])
+      |> Keyword.get(:batch_size)
+
+    table_name =
+      quote do
+        "#{Macro.underscore(__MODULE__)}_embeddings"
+      end
+
+    text_batching =
+      if batch_size do
+        quote do
+          Stream.map(& &1.text)
+          |> Stream.chunk_every(unquote(batch_size))
+        end
+      else
+        quote do
+          Stream.map(& &1.text)
+        end
+      end
+
+    quote do
+      alias Search.{Packages, Repo}
+      use Ecto.Schema
+      import Ecto.{Changeset, Query}
+      alias Pgvector.Ecto.Vector
+
+      def table_name do
+        unquote(table_name)
+      end
+
+      schema unquote(table_name) do
+        field :embedding, Vector
+        belongs_to :doc_fragment, Packages.DocFragment
+
+        timestamps(type: :utc_datetime)
+      end
+
+      def changeset(embedding, attrs) do
+        embedding
+        |> cast(attrs, [:embedding])
+        |> cast_assoc(:doc_fragment, required: true)
+        |> validate_required([:embedding])
+      end
+
+      def embedding_size, do: unquote(embedding_size)
+      def batch_size, do: unquote(batch_size)
+
+      def child_spec(opts) do
+        opts
+        |> Keyword.merge(serving: load_model())
+        |> Nx.Serving.child_spec()
+      end
+
+      @doc """
+      Embeds any doc fragments which do not have an embedding yet. Recieves an optional callback,
+      which is called to notify about the embedding progress with the tuple {total, done} as its argument.
+      """
+      def embed(progress_callback \\ &Function.identity/1) do
```

Comment:
Thanks for the tip regarding macros - I did read the best metaprogramming practices, but to use them is practice is another thing. 

As for whether the macro is even needed here, I opted to create one because AFAIK we plan on benchmarking many models and further on experimenting with finetuning etc. I thought providing a framework for that early on would be a good idea, but after reading your comments I'm not so sure anymore. Could you give some more of your thoughts on this?

Reply:
The modules differ only in configuration, so what we could do is have a single implementation that accepts config. We could define a behaviour to make the contract more explicit.

So the idea is that in the config we would have roughly this:

```elixir
config :search,
  embedding_providers: [
    albert_small:
      {Search.Embeddings.BumblebeeProvider,
        model: {:hf, "sentence-transformers/paraphrase-albert-small-v2"},
        embedding_size: 768,
        serving_opts: [
          compile: [batch_size: 16, sequence_length: 100],
          defn_options: [compiler: EXLA]
        ]},
    paraphrase_l3:
      {Search.Embeddings.BumblebeeProvider,
        model: {:hf, "sentence-transformers/paraphrase-MiniLM-L3-v2"},
        embedding_size: 384,
        serving_opts: [
          compile: [batch_size: 16, sequence_length: 512],
          defn_options: [compiler: EXLA]
        ]}
  ]
```

And then

```elixir
defmodule Search.Embeddings.BumblebeeProvider do
  @behaviour Search.Embeddings.Provider

  @impl true
  def embed(..., config)

  @impl true
  def child_spec(opts, config)
end
```

And finally the API would be like `Search.Embeddings.embed(:albert_small, ...)`, it would fetch the config and call the module.

The main question is how does this relate to the database. Could be a single table with model key, or perhaps we could use a single schema and but use different table name on the fly, based on the provider key.

The benefit of such explicit contract is that plugging something else than Bumblebee would be easy, and this is a good flexibility to have for ML models.

This is just a rough sketch, happy to hear what you guys think!

Reply:
Yup, sounds good to me @jonatanklosko!